staging/src/k8s.io/apiserver/pkg/admission/plugin/finalizerrestriction/admission.go

```diff
@@ -0,0 +1,118 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package finalizerrestriction
+
+import (
+	"fmt"
+	"io"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/klog"
+)
+
+const (
+	// PluginName indicates the name of admission plugin
+	PluginName = "FinalizerRestriction"
+)
+
+// Register registers a plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return &FinalizerRestriction{}, nil
+	})
+}
+
+// FinalizerRestriction is an implementation of admission Interface.
+// It restricts on who can add/remove finalizer on resources.
+type FinalizerRestriction struct {
+	*admission.Handler
+
+	authorizer authorizer.Authorizer
+}
+
+var _ admission.Interface = &FinalizerRestriction{}
+var _ initializer.WantsAuthorizer = &FinalizerRestriction{}
+
+// SetAuthorizer injects the authorizer
+func (r *FinalizerRestriction) SetAuthorizer(authorizer authorizer.Authorizer) {
+	r.authorizer = authorizer
+}
+
+// ValidateInitialization verifies the authorizer injection.
+func (r *FinalizerRestriction) ValidateInitialization() error {
+	if r.authorizer == nil {
+		return fmt.Errorf("missing authorizer")
+	}
+	return nil
+}
+
+// Admit decides whether the finalizer can be added or removed.
+func (r *FinalizerRestriction) Admit(a admission.Attributes, o admission.ObjectInterfaces) error {
+	if a.GetOperation() != admission.Create && a.GetOperation() != admission.Update {
+		return nil
+	}
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	obj, err := meta.Accessor(a.GetObject())
+	if err != nil {
+		klog.V(2).Infof("can not get meta accessor: %s", err)
+		return nil
+	}
+
+	newFinalizers := sets.NewString(obj.GetFinalizers()...)
+	oldFinalizers := sets.NewString()
+
+	if a.GetOldObject() != nil {
+		oldObj, err := meta.Accessor(a.GetOldObject())
+		if err != nil {
+			klog.V(2).Infof("can not get meta accessor for old object: %s", err)
+			return nil
+		}
+		oldFinalizers.Insert(oldObj.GetFinalizers()...)
+	}
+
+	union := newFinalizers.Union(oldFinalizers)
+	intersection := newFinalizers.Intersection(oldFinalizers)
+
+	for _, finalizer := range union.Difference(intersection).List() {
+		attr := authorizer.AttributesRecord{
+			User:            a.GetUserInfo(),
+			Verb:            "finalize",
+			Namespace:       a.GetNamespace(),
+			Name:            finalizer,
```

Comment:
This would need to be the name of the resource, no? I don't think there's a place to pass an argument / subresource specifier, which is really what you need. I am not sure how you can present this in the form of authorizer attributes.

Reply:
There are:

1. Verb:
1. Resource:
1. Name

which I can specify, and I kind of think the combination of:

Finalize(verb) Pods(resource) with Orphan(resourceName)

is a good choice. I guess that we don't care about the pods name(in the example above). But I'm uncertain, maybe there is a case where we want to control the permission to finalize a single object, do we?

Reply:
would it make sense to separate the permission check for updating a particular object's finalizers from the permission check for the ability to add/remove a particular finalizer string (and both would be required to pass)?

For finalizing an object, I'd suggest checking `update` permission on the synthetic `finalizers` subresource of the object being modified, which is the same check as the `OwnerReferencesPermissionEnforcement` admission plugin:

https://github.com/kubernetes/kubernetes/blob/ee84c95bdbec91a49d2d8a666987a7a7632ddc6b/plugin/pkg/admission/gc/gc_admission.go#L189-L203

For each finalizer string being added or removed, I'd suggest checking:
* verb: `create` (when adding) or `delete` (when removing) permission on a set of attributes for each finalizer string being added or removed
* apiGroup: some dns-qualified name specific to your webhook
* resource: `finalizers`
* resourceName: `<specific finalizer string being added or removed>`

That would allow the following types of configurations:
* Authorize a user to manage all finalizers on all objects:
    ```yaml
    - verbs: ["update"]
      apiGroups: ["*"]
      resources: ["*/finalizers"]
    - verbs: ["create","delete"]
      apiGroups: ["yourwebhook.io"]
      resources: ["finalizers"]
    ```
* Authorize a user to manage a specific finalizer on all objects:
    ```yaml
    - verbs: ["update"]
      apiGroups: ["*"]
      resources: ["*/finalizers"]
    - verbs: ["create","delete"]
      apiGroups: ["yourwebhook.io"]
      resources: ["finalizers"]
      resourceNames: ["somefinalizer.io/foo"]
    ```
* Authorize a user to add, but not remove, a specific finalizer on all objects:
    ```yaml
    - verbs: ["update"]
      apiGroups: ["*"]
      resources: ["*/finalizers"]
    - verbs: ["create"]
      apiGroups: ["yourwebhook.io"]
      resources: ["finalizers"]
      resourceNames: ["somefinalizer.io/foo"]
    ```
* Authorize a user to manage all finalizers on a specific resource type:
    ```yaml
    - verbs: ["update"]
      apiGroups: [""]
      resources: ["pods/finalizers"]
    - verbs: ["create","delete"]
      apiGroups: ["yourwebhook.io"]
      resources: ["finalizers"]
    ```
* Authorize a user to manage a specific finalizer on a specific resource type:
    ```yaml
    - verbs: ["update"]
      apiGroups: [""]
      resources: ["pods/finalizers"]
    - verbs: ["create","delete"]
      apiGroups: ["yourwebhook.io"]
      resources: ["finalizers"]
      resourceNames: ["somefinalizer.io/foo"]
    ```
* Authorize a user to add, but not remove, a specific finalizer on a specific resource type:
    ```yaml
    - verbs: ["update"]
      apiGroups: [""]
      resources: ["pods/finalizers"]
    - verbs: ["create"]
      apiGroups: ["yourwebhook.io"]
      resources: ["finalizers"]
      resourceNames: ["somefinalizer.io/foo"]
    ```

Reply:
Thanks for sharing the thoughts. I just have one question: Looks like we are introducing a new semantic of resourceNames with `*/finalizers`(which matches `pods/finalizers`, `services/finalizers`, etc). Is this something which has been agreed upon?

Reply:
> Is this something which has been agreed upon(I mean, is it already supported by current RBAC authroization, let me check anyway)?

Yes. See use for granting the HPA controller access to all scale subresources:

https://github.com/kubernetes/kubernetes/blob/2c2ca27bfc538733c0f0ddbab7d3f2ff8707403d/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/testdata/controller-roles.yaml#L390-L404

Reply:
Great!